Source/Core/Core/PowerPC/JitArm64/JitAsm.cpp

```diff
@@ -27,6 +27,10 @@ void JitArm64::GenerateAsm()
   BitSet32 regs_to_save_fpr(ALL_CALLEE_SAVED_FPR);
   enter_code = GetCodePtr();
 
+#ifdef IPHONEOS
```

Comment:
But why? I know this stuff may be obvious, since you've written it. But out of context, there's no comment to explain why this sort of thing is necessary. My main concern up to this point in reviewing the code is that things do not adequately explain their existence to people that may have to maintain this or work on it that initially don't know much about the iOS ecosystem. I also, for lack of a better expression, feel like a lot of these ifdefs are "bolting" onto the existing API instead of cleanly separating itself out.

I would strongly consider making it more obvious why things these need to exist or trying to factor things out so platform-specifics aren't directly within the codepaths like this.

In many cases even the commit messages don't really provide any explanation about why an iOS-specific bit of code needs to exist (it'll say *what* something is doing, but not *why* it's there). i.e. It should be obvious to even those without intimate knowledge (or any at all) of a particular platform why it needs to do something different.

For example, off the top of my head I can think of to trim this down a bit is to do something like:

```cpp
// Apply relevant name 
template <typename Callable>
void SomeFunction(Callable generator)
{
  // Ifdefs go here, so they're localized, along with a comment explaining why
  // they need to exists.

  // Unprotect memory
  generator();
  // Protect memory
}
```

Then it can (ideally) just be used like:

```cpp
SomeFunction([&] {
  // All the assembly generation goes here.
});
```

This way you don't have all these ifdefs all over the place, removing platform-specifics from the main codepath. It also makes it more difficult to forget to properly handle iOS (and by extension makes it easier to catch these things in a review, since seeing any generation outside of the protection function would likely be an accident).

Probably my main reservation against wanting to see this be merged is that it adds a lot of ifdefs for a single platform all over the place in the AArch64 JIT code and in some other places, in some cases, providing specific behavior for a single platform in an otherwise generic interface, rather than isolating them from everything else (on a similar note to what @Helios747 brought up regarding ifdefs).

Don't get me wrong, I think the idea is cool, but I definitely think it needs a bit of work as well so that we're not making things look kind of sketchy for everything else just for a single platform.

Reply:
Thank you so much for the in-depth explanation. I'll get to work on cleaning everything up.

Reply:
My suggesting here is to replace those ifdef with simple if() cases. So in the end, you might end up with a single ifdef, which enables the mprotect workaround of the Jit. This would allow me to also test it on other plattforms.